x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization2/TokenListCategory.java

```diff
@@ -0,0 +1,754 @@
+    public void mergeWith(TokenListCategory other) {
+        mergeWith(
+            other.maxUnfilteredStringLength,
+            other.baseWeightedTokenIds,
+            other.orderedCommonTokenBeginIndex,
+            other.orderedCommonTokenEndIndex,
+            other.commonUniqueTokenIds,
+            other.numMatches
+        );
+    }
+
+    private void mergeWith(
+        int unfilteredLength,
+        List<TokenAndWeight> weightedTokenIds,
+        int orderedCommonTokenBeginIndex,
+        int orderedCommonTokenEndIndex,
+        List<TokenAndWeight> uniqueTokenIds,
+        long numMatches
+    ) {
+        updateCommonUniqueTokenIds(uniqueTokenIds);
+        updateOrderedCommonTokenIds(weightedTokenIds, orderedCommonTokenBeginIndex, orderedCommonTokenEndIndex);
+        if (unfilteredLength > maxUnfilteredStringLength) {
+            maxUnfilteredStringLength = unfilteredLength;
+        }
+        this.numMatches += numMatches;
+    }
+
+    public void addSubAggs(InternalAggregations aggs) {
+        if (subAggs.isEmpty()) {
+            // Only create the mutable list when needed. Then objects where subAggs
+            // is not used at all can all reference the singleton empty list instead of
+            // needlessly allocating extra memory.
+            subAggs = new ArrayList<>();
+        }
+        subAggs.add(aggs);
+    }
+
+    public List<InternalAggregations> getSubAggs() {
+        return subAggs;
+    }
+
+    /**
+     * Updates the common unique token IDs to remove any that aren't in a new list
+     * of token IDs. Since both lists are sorted the approach is to step through
+     * both lists in parallel looking for differences.
+     */
+    private void updateCommonUniqueTokenIds(List<TokenAndWeight> newUniqueTokenIds) {
+
+        boolean changed = false;
+        int commonIndex = 0;
+        int newIndex = 0;
+
+        while (commonIndex < commonUniqueTokenIds.size()) {
+            TokenAndWeight commonTokenAndWeight = commonUniqueTokenIds.get(commonIndex);
+            if (newIndex >= newUniqueTokenIds.size() || commonTokenAndWeight.getTokenId() < newUniqueTokenIds.get(newIndex).getTokenId()) {
+                commonUniqueTokenWeight -= commonTokenAndWeight.getWeight();
+                commonUniqueTokenIds.remove(commonIndex);
+                changed = true;
+            } else {
+                TokenAndWeight newTokenAndWeight = newUniqueTokenIds.get(newIndex);
+                if (commonTokenAndWeight.getTokenId() == newTokenAndWeight.getTokenId()) {
+                    if (commonTokenAndWeight.getWeight() == newTokenAndWeight.getWeight()) {
+                        ++commonIndex;
+                    } else {
+                        commonUniqueTokenWeight -= commonTokenAndWeight.getWeight();
+                        commonUniqueTokenIds.remove(commonIndex);
+                        changed = true;
+                    }
+                }
+                ++newIndex;
+            }
+        }
```

Comment:
In that StackOverflow answer it looks like the "shift down" method is comparably fast.  It has the benefit that the member variable can still be `final`, which is one less thing to worry about when writing other methods.  I'll change it to use this method.

However, I doubt it will make much difference to the overall timings because what tends to happen is that very quickly the unique tokens get whittled down to the ones that will eventually define the category and then we don't make any further changes.  So the first few merges result in removals but after that there aren't any more.